Makefile

```diff
@@ -96,23 +96,17 @@ define fetch_go_bin_version
 
 endef
 
+# require_clean_work_tree check git tree and exit if any unstanged change.
+# arguments:
+# $(1): Help info. (e.g "Please clean up white noise"). It is optional.
+# The default help info does not need quote or would print quote.
 define require_clean_work_tree
 	@git update-index -q --ignore-submodules --refresh
 
     @if ! git diff-files --quiet --ignore-submodules --; then \
-        echo >&2 "cannot $1: you have unstaged changes."; \
-        git diff-files --name-status -r --ignore-submodules -- >&2; \
-        echo >&2 "Please commit or stash them."; \
+        echo >&2 "$(if $1,$1,You have unstaged changes. Please commit or stash them.)"; \
         exit 1; \
     fi
-
-    @if ! git diff-index --cached --quiet HEAD --ignore-submodules --; then \
```

Comment:
This function was also used previously to ensure a clean git state. Is it really safe to allow `staged` then?

Essentially this function is used for other cases then just check if something was changed in for format/clean whitespaces (:

Reply:
In fact, format or clean whitespaces does not exit 1 but change directly, so git state check works.
With default help info, we can handle most case i think.

Reply:
> In fact, format or clean whitespaces does not exit 1 but change directly, so git state check works.

I know and that's ok. We use this function to exit with 1.

Ok, checked and we are the only users of this function now. This means that it's fine to change it in the way you want but either let's rename it or get back the check for `staged` changes as the function name is misleading now: In this form, the function does not check clean working tree properly.

Reply:
I split them into 2 function, may be more clear.

Reply:
Then let's remove `ensure_no_staged` if we don't use it. (: Why we cannot keep all in same function and just not remove staged one?

Reply:
I think the staged one is more useful at local, no staged change could exist in a pr, so i split them.
I would like to add ensure_no_staged in `make docs` and `make format` if possible.

Reply:
but we never use it  (: